### PR TITLE
feat: add toulligqc wrapper

### PR DESCRIPTION
**Pull request by Salome-Brunon**
Added wrapper for ToulligQC which is a quality control tool for nanopore data.

### QC

* [X] I confirm that I have followed the [documentation for contributing to `snakemake-wrappers`](https://snakemake-wrappers.readthedocs.io/en/stable/contributing.html).

While the contributions guidelines are more extensive, please particularly ensure that:
* [X] `test.py` was updated to call any added or updated example rules in a `Snakefile`
* [X] `input:` and `output:` file paths in the rules can be chosen arbitrarily
* [X] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`)
* [X] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to 
* [X] the `meta.yaml` contains a link to the documentation of the respective tool or command under `url:`
* [X] conda environments use a minimal amount of channels and packages, in recommended ordering

## Summary by CodeRabbit

- **New Features**
	- Introduced a Conda environment configuration for the `toulligqc` project, simplifying setup for users.
	- Added metadata for the `toulligqc` tool, including descriptions and input/output specifications, enhancing user understanding.
	- Implemented multiple workflow rules for processing various sequencing data formats, facilitating integration into bioinformatics pipelines.
	- Developed a wrapper for executing the `toulligqc` command-line tool, improving flexibility in handling different input file types.

- **Tests**
	- Added new test cases for the `toulligqc` process to ensure functionality is covered.

### CODE REVIEW

File: bio/toulligqc/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+__email__ = "salome.brunon@gmail.com"
+__license__ = "CECILL-2.1"
+
+from snakemake.shell import shell
+from pathlib import Path
+
+input_path = Path(snakemake.input[0])
+if input_path.suffix in [".fastq", ".fastq.gz", ".fasta", ".fasta.gz"]:
+    input_flag = "--fastq "
+elif input_path.suffix == ".bam":
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:12]: Does it only support BAM? What about SAM and CRAM?

File: bio/toulligqc/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+name: "toulligqc"
+description: "A post sequencing QC tool for Oxford Nanopore sequencers."
+url: "https://github.com/GenomiqueENS/toulligQC"
+authors:
+  - "Ali Hamraoui"
+  - "Salomé Brunon"
```
Comment (fgvieira) [inline bio/toulligqc/meta.yaml:6]:
```suggestion
  - Ali Hamraoui
  - Salomé Brunon
```

File: bio/toulligqc/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+__author__ = "Salomé Brunon"
+__copyright__ = "Copyright 2024, Salomé Brunon"
+__email__ = "salome.brunon@gmail.com"
+__license__ = "CECILL-2.1"
+
+from snakemake.shell import shell
+from pathlib import Path
+
+input_path = Path(snakemake.input[0])
+if input_path.suffix in [".fastq", ".fastq.gz", ".fasta", ".fasta.gz"]:
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:10]: I think `Path.suffix` of `in.fastq.gz` will return `.gz`.

File: bio/toulligqc/wrapper.py (hunk 1 above)
Comment (Salome-Brunon) [reply bio/toulligqc/wrapper.py:12]: Yes, it only supports BAM and FASTQ files as input!

File: bio/toulligqc/wrapper.py (hunk 3 above)
Comment (Salome-Brunon) [reply bio/toulligqc/wrapper.py:10]: should I revert back to this following code then?

input_flag = "--sequencing-summary-source "
input_path = Path(snakemake.input[0])
if snakemake.input[0].endswith(".fastq") or snakemake.input[0].endswith(".fastq.gz") or snakemake.input[0].endswith(".fasta") or snakemake.input[0].endswith(".fastq.gz"):
if input_path.suffix in [".fastq", ".fastq.gz", ".fasta"]:
    input_flag = "--fastq "
    input_flag = "--fastq "
if snakemake.input[0].endswith(".bam"):
elif input_path.suffix == ".bam":

Comment (fgvieira) [reply bio/toulligqc/wrapper.py:10]: No, just remove `.gz` if it exists
```
if snakemake.input[0].endswith(".gz"):
    input_path = Path(snakemake.input[0].removesuffix(".gz"))
else:
    input_path = Path(snakemake.input[0])
```

File: bio/toulligqc/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    input_flag = "--fastq "
+elif input_path.suffix == ".bam":
+    input_flag = "--bam "
+else:
+    input_flag = "--sequencing-summary-source "
+
+extra = snakemake.params.get("extra", "")
+log = snakemake.log_fmt_shell(stdout=True, stderr=True)
+
+shell("toulligqc {extra} {input_flag} {snakemake.input} {log} -n toulligqc_demo --force")
```
[the remark this answers is not in the document]
Comment (Salome-Brunon) [reply bio/toulligqc/wrapper.py:20]: I am not sure what you mean here, would it be possible to get directed to an example? Thank you.

[the remark this answers is not in the document]
Comment (fgvieira) [reply bio/toulligqc/wrapper.py:20]: It seems that the output directory is hardcoded in `-n toulligqc_demo`. What if the user wants to save the output to another folder?
This folder should be inferred from the output file names.

File: bio/toulligqc/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+name: "toulligqc"
+description: "A post sequencing QC tool for Oxford Nanopore sequencers."
+url: "https://github.com/GenomiqueENS/toulligQC"
```
Comment (fgvieira) [inline bio/toulligqc/meta.yaml:3]:
```suggestion
name: toulligqc
description: A post sequencing QC tool for Oxford Nanopore sequencers.
url: https://github.com/GenomiqueENS/toulligQC
```

File: bio/toulligqc/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+extra = snakemake.params.get("extra", "")
+log = snakemake.log_fmt_shell(stdout=True, stderr=True)
+
+shell("toulligqc {extra} {input_flag} {snakemake.input} {log} {outdir_flag} --force")
+
+
+
+
+
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:33]:
```suggestion
shell("toulligqc {extra} {input_flag} {snakemake.input} {log} {outdir_flag} --force")
```

File: bio/toulligqc/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+__author__ = "Salomé Brunon"
+__copyright__ = "Copyright 2024, Salomé Brunon"
+__email__ = "salome.brunon@gmail.com"
+__license__ = "CECILL-2.1"
+
+import os
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:6]:
```suggestion
```

File: bio/toulligqc/wrapper.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+extra = snakemake.params.get("extra", "")
+log = snakemake.log_fmt_shell(stdout=True, stderr=True)
+
+shell("toulligqc {extra} {input_flag} {snakemake.input} {log} {outdir_flag} --force")
+
+
+
+
+
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:32]:
```suggestion
if input_path.suffix in [".fastq", ".fq"]:
    input_flag = "--fastq "
elif input_path.suffix == ".bam":
    input_flag = "--bam "
else:
    input_flag = "--sequencing-summary-source "

outdir = Path(snakemake.output[0]).parent

shell("toulligqc {extra} {input_flag} {snakemake.input} -n {outdir} {log}")
```

File: bio/toulligqc/wrapper.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+__email__ = "salome.brunon@gmail.com"
+__license__ = "CECILL-2.1"
+
+from snakemake.shell import shell
+from pathlib import Path
+
+if snakemake.input[0].endswith(".gz"):
+    input_path = Path(snakemake.input[0].removesuffix(".gz"))
+else:
+    input_path = Path(snakemake.input[0])
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:12]:
```suggestion
from snakemake.shell import shell
from pathlib import Path

extra = snakemake.params.get("extra", "")
log = snakemake.log_fmt_shell(stdout=True, stderr=True)

input_path = Path(snakemake.input[0])
if input_path.suffix == ".gz":
    input_path = input_path.with_suffix("")
```

File: bio/toulligqc/wrapper.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+if input_path.suffix in [".fastq", ".fq"]:
+    input_flag = "--fastq "
+elif input_path.suffix == ".bam":
+    input_flag = "--bam "
+else:
+    input_flag = "--sequencing-summary-source "
+
+outdir = Path(snakemake.output[0]).parent
+
+shell("toulligqc {input_flag} {snakemake.input[0]} {extra} -n {outdir} {log}")
```
Comment (fgvieira) [inline bio/toulligqc/wrapper.py:25]:
```suggestion
shell("toulligqc {input_flag} {snakemake.input[0]} {extra} --force -n {outdir} {log}")
```

**Comment by Salome-Brunon** [comment]
@fgvieira Thank you for all the edits!

**Comment by fgvieira** [comment]
Sure. 😄 
Can you just add some small tests for fastq and bam files?

**Comment by Salome-Brunon** [comment]
> Sure. 😄 Can you just add some small tests for fastq and bam files?

Yes of course, I'll work on that next week!

File: bio/toulligqc/test/Snakefile [hunk 11]
```diff
@@ -0,0 +1,35 @@
+rule toulligqc_sequencing_summary:
+    input:
+        "demodata/sequencing_summary.txt",
+    output:
+        "toulligqc_sequencing_summary/report.html",
+    log:
+        "merged.log",
```
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:7]:
```suggestion
        "toulligqc_sequencing_summary/merged.log",
```

File: bio/toulligqc/test/Snakefile [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    wrapper:
+        "master/bio/toulligqc"
+
+rule toulligqc_bam:
+    input:
+        "demodata/test.sorted.bam",
+    output:
+        "toulligqc_bam/report.html",
+    log:
+        "merged.log",
```
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:19]:
```suggestion
        "toulligqc_bam/merged.log",
```

File: bio/toulligqc/test/Snakefile [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    wrapper:
+        "master/bio/toulligqc"
+
+rule toulligqc_fastq:
+    input:
+        "demodata/test.fastq.gz",
+    output:
+        "toulligqc_fastq/report.html",
+    log:
+        "merged.log",
```
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:31]:
```suggestion
        "toulligqc_fastq/merged.log",
```

**Comment by Salome-Brunon** [comment]
@fgvieira Hello! I have submitted the test bam and fast files, is there anything else needed for this PR to be approved? :)

**Comment by fgvieira** [comment]
you need to add the tests to the `test.py` file. I've suggested the changes, so just accept them all.

File: bio/toulligqc/test/Snakefile (hunk 13 above)
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:31]:
```suggestion
        "toulligqc_fastq/merged.log",
```

File: bio/toulligqc/test/Snakefile (hunk 11 above)
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:7]:
```suggestion
        "toulligqc_sequencing_summary/merged.log",
```

File: bio/toulligqc/test/Snakefile [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    wrapper:
+        "master/bio/toulligqc"
+
+rule toulligqc_bam:
+    input:
+        "demodata/test.sorted.bam",
+    output:
+        "toulligqc_bam/report.html",
+    log:
+        "merged.log",
```
Comment (fgvieira) [inline bio/toulligqc/test/Snakefile:19]:
```suggestion
        "toulligqc_bam/merged.log",
```

**Pull request merged**